tools/autograd/gen_variable_type.py

```diff
@@ -618,9 +642,13 @@ def save_variables(
                 expr = f'{expr}.has_value() ? c10::optional<std::string>(std::string({expr}.value())) : c10::nullopt'
             guard = guard_for(arg)
             if guard is None:
+                if stmts_prepend:
+                    stmts.append(f'{stmts_prepend};')
                 stmts.append(f'grad_fn->{name} = {expr};')
             else:
                 stmts.append(f'if ({guard}) {{')
+                if stmts_prepend:
+                    stmts.append(f'  {stmts_prepend};')
```

Comment:
I'm not familiar with this code -- what is an example of the `guard` here?

Reply:
These guards are used when there are multiple formulas for a given function so that we can avoid saving Tensors when that input doesn't require gradients.
For example, for addmm:
```c++
    if (grad_fn->should_compute_output(1)) {
      grad_fn->mat2_ = SavedVariable(mat2, false);
    }
    grad_fn->alpha = alpha;
    if (grad_fn->should_compute_output(2)) {
      grad_fn->mat1_ = SavedVariable(mat1, false);
    }
```